[16.0][ADD] stock_average_daily_sale

rousseldenis opened the pull request:
Depends on:

- https://github.com/OCA/product-attribute/pull/1192
- https://github.com/OCA/wms/pull/470
- https://github.com/OCA/wms/pull/496
- https://github.com/OCA/product-attribute/pull/1227
- https://github.com/OCA/stock-logistics-warehouse/pull/1580


I've put a comment there : https://github.com/OCA/wms/pull/496#pullrequestreview-1247246210

Because the dependency on stock_storage_type should be avoided

rousseldenis committed:
[16.0][ADD] stock_average_daily_sale

sbejaoui reviewed (approved):
LGTM, thanks

lmignon commented on stock_average_daily_sale/tests/test_average_daily_sale.py:
```diff
@@ -0,0 +1,180 @@
+from dateutil.relativedelta import relativedelta
```
@rousseldenis The hearder is missing...

rousseldenis replied:
Done

lmignon reviewed (approved):
LGTM (Code review + functiional)

rousseldenis committed:
[IMP] stock_average_daily_sale: Store the configurations on profile level + add demo

The configurations are now stored on abc profiles.

Some demo data are loaded in order to show this module features more easily.

rousseldenis committed:
[IMP] stock_average_daily_sale: Improve profile view

simahawk commented:
/ocabot migration stock_average_daily_sale

jbaudoux commented on stock_average_daily_sale/models/stock_average_daily_sale.py:
```diff
@@ -0,0 +1,336 @@
+    )
+    average_daily_sales_count = fields.Float(
+        required=True, digits="Product Unit of Measure"
+    )
```
```suggestion
        help="How much deliveries on average for this product on the period. The spikes are excluded from the average computation."
    )
```

jbaudoux commented on stock_average_daily_sale/models/stock_average_daily_sale.py:
```diff
@@ -0,0 +1,336 @@
+        required=True, digits="Product Unit of Measure"
+    )
+    average_qty_by_sale = fields.Float(
+        required=True, digits="Product Unit of Measure", help="Average Daily Sales Qty"
```
```suggestion
        required=True, digits="Product Unit of Measure", help="The quantity delivered on average for one delivery of this product on the period. The spikes are excluded from the average
/ computation."
```

jbaudoux commented on stock_average_daily_sale/models/stock_average_daily_sale.py:
```diff
@@ -0,0 +1,336 @@
+    average_daily_qty = fields.Float(
+        digits="Product Unit of Measure",
+        required=True,
+        help="The average daily qty sold",
```
```suggestion
        help="The quantity delivered on average on one day for this product on the period. The spikes are excluded from the average computation.",
```

jbaudoux commented on stock_average_daily_sale/models/stock_average_daily_sale.py:
```diff
@@ -0,0 +1,336 @@
+        store=True,
+        index=True,
+    )
+    nbr_sales = fields.Integer(string="Number of Sales", required=True)
```
```suggestion
    nbr_sales = fields.Integer(string="Number of Sales", required=True,
    help="The total amount of deliveries for this product over the complete period")
```

jbaudoux commented on stock_average_daily_sale/models/stock_average_daily_sale.py:
```diff
@@ -0,0 +1,336 @@
+    safety = fields.Float(
+        required=True,
+        help="daily standard deviation * safety factor * sqrt(nbr days into period "
+        "without saturday and sunday",
```
```suggestion
        help="Safety stock to cover the variability of the quantity delivered each day. Formula: daily standard deviation * safety factor * sqrt(nbr days in the period)"
```

jbaudoux commented on stock_average_daily_sale/models/stock_average_daily_sale.py:
```diff
@@ -0,0 +1,336 @@
+        digits="Product Unit of Measure",
+        help="Minimal safety qty into a bin location computed as: "
+        "average daily qty * number days in stock * safety",
+    )
```
It is **+** safety instead of ***** safety.
I would rename this as the **recommended_qty**
```suggestion
    recommended_qty = fields.Float(
        required=True,
        digits="Product Unit of Measure",
        help="Minimal recommended quantity in stock. Formula: average daily qty * number days in stock + safety",
    )
```

jbaudoux commented on stock_average_daily_sale/models/stock_average_daily_sale.py:
```diff
@@ -0,0 +1,336 @@
+        help="Minimal value for the safety qty. Computed as: "
+        "number days in stock * GREATEST(average daily sales count, 1) * "
+        "(average qty by sale + (stddev * safety factor))",
+    )
```
To drop

jbaudoux commented on stock_average_daily_sale/models/stock_average_daily_sale.py:
```diff
@@ -0,0 +1,336 @@
+    warehouse_id = fields.Many2one(comodel_name="stock.warehouse", required=True)
+    zone_location_id = fields.Many2one(
+        string="Location Zone", comodel_name="stock.location", index=True
+    )
```
To drop everywhere as well as the dependency. This brings nothing as we consider moves from output. Zones are in stock, not on output.

jbaudoux commented on stock_average_daily_sale/__manifest__.py:
```diff
@@ -0,0 +1,35 @@
+        Allows to gather delivered products average on daily basis""",
+    "version": "16.0.1.0.0",
+    "license": "AGPL-3",
+    "author": "ACSONE SA/NV,Odoo Community Association (OCA)",
```
For the work on the conception
```suggestion
    "author": "ACSONE SA/NV,BCIM,Odoo Community Association (OCA)",
```

jbaudoux commented on stock_average_daily_sale/readme/CONTRIBUTORS.rst:
```diff
@@ -0,0 +1,2 @@
+* Laurent Mignon <laurent.mignon@acsone.eu>
+* Denis Roussel <denis.roussel@acsone.eu>
```
```suggestion
* Denis Roussel <denis.roussel@acsone.eu>
* Jacques-Etienne Baudoux (BCIM) <je@bcim.be>
```

rousseldenis committed:
[IMP] Remove zone location dependency, rely on warehouse only

rousseldenis committed:
[IMP] stock_average_daily_sale: Improve tests

rousseldenis committed:
[IMP] stock_average_daily_sale: Improve module helps / change field name to recommended_qty

lmignon committed:
[FIX] stock_average_daily_sale: Don't exclude moves on priority

By default the priority in odoo is set to '0'. Before this change
the daily sale was never computed for moves done in a normal process.
There is no need to filter out moves based on the priority from
the computation.

rousseldenis committed:
[IMP] stock_average_daily_sale: Don't use Odoo demo data

sbejaoui committed:
[IMP] stock_average_daily_sale: add contributor

Co-authored-by: Jacques-Etienne Baudoux <je@bcim.be>

sbejaoui commented:
@jbaudoux , can you update your review please

rousseldenis committed:
[IMP] stock_average_daily_sale: Avoid errors in all searches and return browse record

rousseldenis commented:
@jbaudoux Could you update your review ?

jbaudoux reviewed (approved):
Thanks for the fixes. LGTM

What about adding the proposed help?

Small remark: Some commit messages have an overflowing first line. One commit is missing the concerned module.

rousseldenis commented:
> What about adding the proposed help?

If for average_daily_sales_count, it is already there.

rousseldenis committed:
[FIX] stock_average_daily_sale: Use a new cursor to avoid closed one

rousseldenis committed:
[FIX] stock_average_daily_sale: Don't test the not refreshed materialized view

rousseldenis committed:
[FIX] stock_average_daily_sale: Don't check the view existence in tests

As tests create a Savepoint, there is a concurrent query when checking the
view existence.

rousseldenis committed:
[FIX] stock_average_daily_sale: Don't rely on tuples as id

As for x reason, some products appear more than one time in the report,
id generated by the concatenation of product_id and warehouse_id is
irrelevant (as duplicate values possible). Use row_number() instead

lmignon committed:
[IMP] stock_average_daily_sale: avoid concurrent update

rousseldenis commented on stock_average_daily_sale/models/stock_average_daily_sale.py:
```diff
@@ -114,7 +127,7 @@ def _check_view(self):
             raise e
         finally:
-            new_self.env.cr.close()
+            cr.close()
```
@lmignon Is this really wanted (the code before was closing the new created cursor, not the current one)?

lmignon replied:
The code close the new cursor...

```
cr = registry(self._cr.dbname).cursor()
....

cr.close()
```

I'll clean the code to improve readability

```python
    @api.model
    def _check_view(self):
        cr = registry(self._cr.dbname).cursor()
        with closing(cr):
            try:
                return self._check_materialize_view_populated(cr)
            except ObjectNotInPrerequisiteState:
                _logger.warning(
                    _("The materialized view has not been populated. Launch the cron.")
                )
                return False
            except Exception as e:
                raise e

```

dreispt commented:
/ocabot merge nobump

jbaudoux commented:
/ocabot merge nobump
